homeassistant/components/sensor/yr.py

```diff
@@ -120,78 +124,95 @@ def unit_of_measurement(self):
-                self._state = loc_data[self.type]['@mps']
-                break
-            elif self.type == 'windDirection':
-                self._state = float(loc_data[self.type]['@deg'])
-                break
-            elif self.type in ('fog', 'cloudiness', 'lowClouds',
-                               'mediumClouds', 'highClouds'):
-                self._state = loc_data[self.type]['@percent']
-                break
-
 
 class YrData(object):
     """Get the latest data and updates the states."""
 
-    def __init__(self, coordinates):
+    def __init__(self, hass, coordinates, devices):
         """Initialize the data object."""
         self._url = 'http://api.yr.no/weatherapi/locationforecast/1.9/?' \
             'lat={lat};lon={lon};msl={msl}'.format(**coordinates)
-
         self._nextrun = None
+        self.devices = devices
         self.data = {}
-        self.update()
+        self.hass = hass
 
-    def update(self):
+    @asyncio.coroutine
+    def async_update(self):
         """Get the latest data from yr.no."""
-        # Check if new will be available
-        if self._nextrun is not None and dt_util.utcnow() <= self._nextrun:
-            return
+        def try_again(err: str):
+            """Schedule again later."""
+            _LOGGER.warning('Retrying in 15 minutes: %s', err)
+            nxt = dt_util.utcnow() + timedelta(minutes=15)
+            async_track_point_in_utc_time(self.hass, self.async_update, nxt)
+
+        try:
+            with async_timeout.timeout(10, loop=self.hass.loop):
+                resp = yield from self.hass.websession.get(
+                    self._url, timeout=30)
+            if resp.status != 200:
+                return try_again('{} returned {}'
+                                 .format(self._url, resp.status))
+            text = yield from resp.text()
+            self.hass.loop.create_task(resp.release())
+        except asyncio.TimeoutError as err:
+            return try_again(err)
+        except HTTPException as err:
+            resp.close()
+            return try_again(err)
+
         try:
-            with requests.Session() as sess:
-                response = sess.get(self._url)
-        except requests.RequestException:
-            return
-        if response.status_code != 200:
-            return
-        data = response.text
-
-        import xmltodict
-        self.data = xmltodict.parse(data)['weatherdata']
-        model = self.data['meta']['model']
-        if '@nextrun' not in model:
-            model = model[0]
-        self._nextrun = dt_util.parse_datetime(model['@nextrun'])
+            import xmltodict
+            self.data = xmltodict.parse(text)['weatherdata']
+            model = self.data['meta']['model']
+            if '@nextrun' not in model:
+                model = model[0]
+            next_run = dt_util.parse_datetime(model['@nextrun'])
+        except (ExpatError, IndexError) as err:
+            return try_again(err)
```

Comment:
Yes, split it over two lines. If people should not use the return value, it should have none.